cluster/src/shard_lock_manager.rs

```diff
@@ -0,0 +1,149 @@
+            shard_leader_key: SHARD_LEADER_PATH.to_string(),
+            lease_ttl_sec: DEFAULT_LEASE_SECS,
+            etcd_cli: etcd_client,
+            shard_table_cache,
+        };
+    }
+
+    /// Return true if grant lock success, otherwise return false.
+    /// A new thread is started internally and keeps calling keep alive.
+    pub async fn grant_lock(&self, shard_id: u32) -> Result<bool, Error> {
+        // generate key path
+        let key_path = vec![
+            self.root_path.clone(),
+            self.shard_leader_key.clone(),
+            shard_id.to_string(),
+        ]
+        .join("/");
+        info!("generate shard key path {}", key_path);
+
+        // check lock owner, if the locked has been granted, .
+        let get_resp = self.etcd_cli.clone().get(key_path.clone(), None).await;
+        match get_resp {
+            Ok(resp) => {
+                if !resp.kvs().is_empty() {
+                    let value = resp.kvs().get(0).unwrap().value();
+                    info!("lease exists, {:?}", resp);
+                    match String::from_utf8(value.to_vec()) {
+                        Ok(s) => return Ok(s == self.endpoint),
+                        Err(e) => {}
+                    }
+                }
+            }
+            Err(e) => return Err(e),
+        }
+
+        // get lock, put data to etcd with lease.
+        let grant_lease_resp = self
+            .etcd_cli
+            .clone()
+            .lease_grant(self.lease_ttl_sec, None)
+            .await;
+        let lease_id = grant_lease_resp?.id();
+
+        let endpoint = &self.endpoint;
+        let txn_res = self
+            .etcd_cli
+            .clone()
+            .txn(
+                Txn::new()
+                    .when(vec![Compare::version(
+                        key_path.clone(),
+                        CompareOp::Equal,
+                        0,
+                    )])
+                    .and_then(vec![TxnOp::put(
+                        key_path.clone(),
+                        self.endpoint.clone(),
+                        Some(PutOptions::new().with_lease(lease_id).with_prev_key()),
+                    )]),
+            )
+            .await;
+
+        let put_resp = txn_res?.op_responses().get(0);
+
+        let mut keep_alive_interval =
+            tokio::time::interval(Duration::from_secs(KEEP_ALIVE_PERIOD_SECS));
+
+        let mut client = self.etcd_cli.clone();
+        let shard_tables_cache = self.shard_table_cache.clone();
+        tokio::spawn(async move {
+            let (mut keeper, mut stream) = client.lease_keep_alive(lease_id).await.unwrap();
+            loop {
+                keep_alive_interval.tick().await;
+
+                info!("lease {:?} keep alive start", lease_id);
+                let resp = keeper.keep_alive().await;
+                match resp {
+                    Ok(_) => {}
+                    Err(e) => {
+                        // Close Shard
```

Comment:
More things besides remove shard from the cache need to be done when the lock lease is expired. Add a TODO here.